src/utils/polygonUtils.cpp

```diff
@@ -994,5 +994,44 @@ bool PolygonUtils::polygonCollidesWithLineSegment(const Polygons& polys, const P
     return polygonCollidesWithLineSegment(polys, transformed_startPoint, transformed_endPoint, transformation_matrix);
 }
 
+bool PolygonUtils::polygonsIntersect(const Polygons& poly_a, const Polygons &poly_b)
+{
+    // only do the full intersection when the polys' BBs overlap
+    AABB bba(poly_a);
+    AABB bbb(poly_b);
+    return bba.hit(bbb) && poly_a.intersection(poly_b).size() > 0;
+}
+
+bool PolygonUtils::polygonOutlinesAdjacent(const ConstPolygonRef inner_poly, const ConstPolygonRef outer_poly, const coord_t max_gap)
+{
+    const coord_t max_gap2 = max_gap * max_gap;
+    const unsigned outer_poly_size = outer_poly.size();
+    for (unsigned line_index = 0; line_index < outer_poly_size; ++line_index)
+    {
+        const Point lp0 = outer_poly[line_index];
+        const Point lp1 = outer_poly[(line_index + 1) % outer_poly_size];
+        for (Point inner_poly_point : inner_poly)
+        {
+            if (LinearAlg2D::getDist2FromLineSegment(lp0, inner_poly_point, lp1) < max_gap2)
```

Comment:
That's quadratic! Quite expensive, especially when combined with the additional quadratic factor in `processInsets()`. I can think of one optimisation that would save a significant amount of time here in the random case: Add a heuristic by first checking whether the AABB's of the polygons are adjacent.

Off the top of my head I can also think of an `O(n log n)` algorithm: Find the MST of the combined set of vertices (`O(n log n)`), then check each of those edges (`O(n)`) for the smallest connection of a point from one subset to the other. But perhaps that is too much work and in many small-scale slices it would perform slower. Perhaps the heuristic combined with brute-force search is the best we can achieve.